src/FEValues/face_values.jl

```diff
@@ -31,143 +31,57 @@ values of nodal functions, gradients and divergences of nodal functions etc. on
-    qr::QR
+struct FaceValues{FV, GV, QR, detT, nT, V_FV<:AbstractVector{FV}, V_GV<:AbstractVector{GV}} <: AbstractFaceValues
+    fun_values::V_FV # AbstractVector{FunctionValues}
+    geo_values::V_GV # AbstractVector{GeometryValues}
+    qr::QR           # FaceQuadratureRule
+    detJdV::detT     # AbstractVector{<:Number}
+    normals::nT      # AbstractVector{<:Vec}
     current_face::ScalarWrapper{Int}
-    func_interp::IP
-    geo_interp::GIP
 end
 
-# Common initializer code for constructing FaceValues after the types have been determined
-function FaceValues{IP, N_t, dNdx_t, dNdξ_t, T, dMdξ_t, QR, Normal_t, GIP}(qr::QR, ip::IP, gip::GIP) where {
-    IP, N_t, dNdx_t, dNdξ_t, T, dMdξ_t, QR, Normal_t, GIP,
-}
-    @assert isconcretetype(IP)     && isconcretetype(N_t)      && isconcretetype(dNdx_t) &&
-            isconcretetype(dNdξ_t) && isconcretetype(T)        && isconcretetype(dMdξ_t) &&
-            isconcretetype(QR)     && isconcretetype(Normal_t) && isconcretetype(GIP)
-    # Quadrature
-    max_n_qpoints = maximum(getnquadpoints, qr.face_rules; init = 0)
-    n_faces = length(qr.face_rules)
-
-    # Normals
-    normals = zeros(Normal_t, max_n_qpoints)
-
-    # Field interpolation
-    n_func_basefuncs = getnbasefunctions(ip)
-    N    = fill(zero(N_t)    * T(NaN), n_func_basefuncs, max_n_qpoints, n_faces)
-    dNdx = fill(zero(dNdx_t) * T(NaN), n_func_basefuncs, max_n_qpoints, n_faces)
-    dNdξ = fill(zero(dNdξ_t) * T(NaN), n_func_basefuncs, max_n_qpoints, n_faces)
-
-    # Geometry interpolation
-    n_geom_basefuncs = getnbasefunctions(gip)
-    M    = fill(zero(T)      * T(NaN), n_geom_basefuncs, max_n_qpoints, n_faces)
-    dMdξ = fill(zero(dMdξ_t) * T(NaN), n_geom_basefuncs, max_n_qpoints, n_faces)
-
-    for face in 1:n_faces, (qp, ξ) in pairs(getpoints(qr, face))
-        for basefunc in 1:n_func_basefuncs
-            dNdξ[basefunc, qp, face], N[basefunc, qp, face] = shape_gradient_and_value(ip, ξ, basefunc)
-        end
-        for basefunc in 1:n_geom_basefuncs
-            dMdξ[basefunc, qp, face], M[basefunc, qp, face] = shape_gradient_and_value(gip, ξ, basefunc)
-        end
-    end
-
-    detJdV = fill(T(NaN), max_n_qpoints, n_faces)
-
-    FaceValues{IP, N_t, dNdx_t, dNdξ_t, T, dMdξ_t, QR, Normal_t, GIP}(N, dNdx, dNdξ, detJdV, normals, M, dMdξ, qr, ScalarWrapper(0), ip, gip)
-end
-
-# Common entry point that fills in the numeric type and geometric interpolation
-function FaceValues(qr::FaceQuadratureRule, ip::Interpolation,
-        gip::Interpolation = default_geometric_interpolation(ip))
-    return FaceValues(Float64, qr, ip, gip)
+function FaceValues(::Type{T}, fqr::FaceQuadratureRule, ip_fun::Interpolation, ip_geo::VectorizedInterpolation{sdim}=default_geometric_interpolation(ip_fun)) where {T,sdim} 
+    geo_values = [GeometryValues(T, ip_geo.ip, qr, RequiresHessian(ip_fun, ip_geo)) for qr in fqr.face_rules]
+    fun_values = [FunctionValues(T, ip_fun, qr, ip_geo) for qr in fqr.face_rules]
```

Comment:
They could be, but I don't think it makes a difference since they contain values with mutable values. In that case, I would probably go for tuple, but it would add an additional type parameter. 
And I think this should match the way it is done in `FaceQuadratureRule`